src/modules/avformat/producer_avformat.c

```diff
@@ -1658,6 +1681,53 @@ static int ignore_send_packet_result(int result)
 	return result >= 0 || result == AVERROR(EAGAIN) || result == AVERROR_EOF || result == AVERROR_INVALIDDATA || result == AVERROR(EINVAL);
 }
 
+static void* packets_worker( void *param )
+{
+	producer_avformat self = param;
+	AVPacket *pkt = av_packet_alloc();
+	if ( !pkt ) {
+		mlt_log_error( MLT_PRODUCER_SERVICE(self->parent), "av_packet_alloc failed\n" );
+		exit( EXIT_FAILURE );
+	}
+
+	for (;;) {
+		pthread_mutex_lock( &self->packets_mutex );
+
+	check_stop:
+		if ( self->packets_thread_stop ) {
+			av_packet_free( &pkt );
+			pthread_mutex_unlock( &self->packets_mutex );
+			return NULL;
+		}
+
+		if ( mlt_deque_count( self->vpackets ) >= 1 ) {
+			pthread_cond_wait( &self->packets_cond, &self->packets_mutex );
+			goto check_stop;
+		}
+
+		int ret = av_read_frame( self->video_format, pkt );
+		self->packets_thread_ret = ret;
+
+		// don't bother signalling on EAGAIN
+		if ( ret != AVERROR(EAGAIN) ) {
+			if ( ret >= 0 ) {
+				if ( pkt->stream_index == self->video_index ) {
+					mlt_deque_push_back( self->vpackets, av_packet_clone( pkt ) );
+				} else if ( !self->video_seekable && pkt->stream_index == self->audio_index ) {
+					mlt_deque_push_back( self->apackets, av_packet_clone( pkt ) );
+				}
+			} else {
+				mlt_log_verbose( MLT_PRODUCER_SERVICE(self->parent), "av_read_frame returned error %d inside packets_worker\n", ret );
+			}
+
+			pthread_cond_signal( &self->packets_cond );
+		}
+
+		av_packet_unref( pkt );
+		pthread_mutex_unlock( &self->packets_mutex );
```

Comment:
Micro-optimization: the mutex is unlocked here only to be immediately locked at the top of the loop. Suggest to move the lock/unlock outside of the loop to avoid this.

Reply:
Good idea. pthread_cond_wait() will unlock the mutex when appropriate, and it will be finally unlocked on return. I'll make the change in a second version of this PR.